sycl/include/CL/sycl/aliases.hpp

```diff
@@ -0,0 +1,115 @@
+  MAKE_VECTOR_ALIAS(long, long, N)                                             \
+  MAKE_VECTOR_ALIAS(float, float, N)                                           \
+  MAKE_VECTOR_ALIAS(double, double, N)                                         \
+  MAKE_VECTOR_ALIAS(half, ::half, N)
+
+#define MAKE_VECTOR_ALIASES_FOR_OPENCL_TYPES(N)                                \
+  MAKE_VECTOR_ALIAS(cl_char, cl::sycl::cl_char, N)                             \
+  MAKE_VECTOR_ALIAS(cl_uchar, cl::sycl::cl_uchar, N)                           \
+  MAKE_VECTOR_ALIAS(cl_short, cl::sycl::cl_short, N)                           \
+  MAKE_VECTOR_ALIAS(cl_ushort, cl::sycl::cl_ushort, N)                         \
+  MAKE_VECTOR_ALIAS(cl_int, cl::sycl::cl_int, N)                               \
+  MAKE_VECTOR_ALIAS(cl_uint, cl::sycl::cl_uint, N)                             \
+  MAKE_VECTOR_ALIAS(cl_long, cl::sycl::cl_long, N)                             \
+  MAKE_VECTOR_ALIAS(cl_ulong, cl::sycl::cl_ulong, N)                           \
+  MAKE_VECTOR_ALIAS(cl_float, cl::sycl::cl_float, N)                           \
+  MAKE_VECTOR_ALIAS(cl_double, cl::sycl::cl_double, N)                         \
+  MAKE_VECTOR_ALIAS(cl_half, cl::sycl::half, N)
+
+#define MAKE_VECTOR_ALIASES_FOR_SIGNED_AND_UNSIGNED_TYPES(N)                   \
+  MAKE_VECTOR_ALIAS(schar, signed char, N)                                     \
+  MAKE_VECTOR_ALIAS(uchar, unsigned char, N)                                   \
+  MAKE_VECTOR_ALIAS(ushort, unsigned short, N)                                 \
+  MAKE_VECTOR_ALIAS(uint, unsigned int, N)                                     \
+  MAKE_VECTOR_ALIAS(ulong, unsigned long, N)                                   \
+  MAKE_VECTOR_ALIAS(longlong, long long, N)                                    \
+  MAKE_VECTOR_ALIAS(ulonglong, unsigned long long, N)
+
+#define MAKE_VECTOR_ALIASES_FOR_VECTOR_LENGTH(N)                               \
+  MAKE_VECTOR_ALIASES_FOR_ARITHMETIC_TYPES(N)                                  \
+  MAKE_VECTOR_ALIASES_FOR_OPENCL_TYPES(N)                                      \
+  MAKE_VECTOR_ALIASES_FOR_SIGNED_AND_UNSIGNED_TYPES(N)
+
+namespace cl {
+namespace sycl {
+namespace detail {
+using conditional_data_t = decltype(1 == 1);
+} // namespace detail
+using byte = ::uint8_t;
+using half = ::half;
+using schar = signed char;
+using uchar = unsigned char;
+using ushort = unsigned short;
+using uint = unsigned int;
+using ulong = unsigned long;
+using longlong = long long;
+using ulonglong = unsigned long long;
+using cl_bool = detail::conditional_data_t;
+using cl_char = ::int8_t;
+using cl_uchar = ::uint8_t;
+using cl_short = ::int16_t;
+using cl_ushort = ::uint16_t;
+using cl_int = ::int32_t;
+using cl_uint = ::uint32_t;
+using cl_long = ::int64_t;
+using cl_ulong = ::uint64_t;
+using cl_half = ::half;
+using cl_float = float;
+using cl_double = double;
+using cl_size_t = std::size_t;
+using cl_ptrdiff_t = std::ptrdiff_t;
```

Comment:
> Some types in SYCL vary according to pointer size or vary on the host according to the host ABI, such as size_t
> or long. In order for the the SYCL device compiler to ensure that the sizes of these types match the sizes on the
> host and to enable data of these types to be shared between host and device, the OpenCL interoperability types
> are defined, cl::sycl::cl_int and **cl::sycl::cl_size_t**
Probably just these types are not described, although it should be.

Reply:
@romanovvlad You found it?

Reply:
Yes, it explains why you defining cl_size_t but what about cl_ptrdiff_t ?

Reply:
I believe that SYCL spec missed them definition in the scalar types table. I will create an issue about it. Do you think adding this is wrong now?

Reply:
Thanks!

Reply:
> I believe that SYCL spec missed them definition in the scalar types table. I will create an issue about it.

I am unsure about `cl::sycl::cl_size_t` and `cl::sycl::cl_ptrdiff_t`.
How could it work if you have both 32-bit and 64-bit OpenCL devices at the same time in your system?

Reply:
I, actually, have doubts that SYCL spec define **any** aliases to OpenCL types at all.

Section 3.10 Language restriction in kernels:

> Some types in SYCL vary according to pointer size or vary on the host according to the host ABI, such as `size_t` or `long`. In order for the the SYCL device compiler to ensure that the sizes of these types match the sizes on the host and to enable data of these types to be shared between host and device, the OpenCL interoperability types are defined, `cl::sycl::cl_int` and `cl::sycl::cl_size_t`.

However, Section 4.10.1 Scalar data types:

> The OpenCL C language standard [1, §6.11] defines its own built-in scalar data types, and these have additionalrequirements in  terms of  size and  signedness on top  of what  is guaranteed  by ISO C++. For the purpose of interoperability and portability, SYCL defines a set of **aliases to C++ types** within the `cl::sycl` namespace using the `cl_` prefix. These aliases are described in Table 4.94.

Please note **aliases to C++ types**. Yes, Table 4.94: Scalar data type aliases supported by SYCL restrict sizes of such data types and makes them the same as in OpenCL, but for example, `cl_size_t` is not listed in there.

Having that said I would rather remove `cl_size_t` and `cl_ptrdiff_t` aliases until we have clarification from the spec side. Users still could use `size_t` which will correspond to a host-side `size_t`.

One more note here: Table 6.1: Fundamental data types supported by SYCL doesn't contain `ptrdiff_t`

@mkinsner, @jbrodman, @Pennycook: FYI

Reply:
Sorry, I missed that issue.
I agree with @AlexeySachkov and @keryell that `cl_size_t` and `cl_ptrdiff_t` doesn't make sense (as well as `cl_intptr_t` and `cl_uintptr_t`).
@alexeyvoronov, please, remove them.

Reply:
@bader then why the specification has cl_size_t ? it's mistake?

Reply:
I think so.

I've found only on mention of cl_size_t in the spec.

> Some types in SYCL vary according to pointer size or vary on the host according to the host ABI, such as size_t or long. In order for the the SYCL device compiler to ensure that the sizes of these types match the sizes on the host and to enable data of these types to be shared between host and device, the OpenCL interoperability types are defined, cl::sycl::cl_int and cl::sycl::cl_size_t.

The first part of this sentence, seems wrong to me:
> **_In order for the the SYCL device compiler to ensure that the sizes of these types match the sizes on the host_** and to enable data of these types to be shared between host and device, the OpenCL interoperability types are defined, cl::sycl::cl_int and cl::sycl::cl_size_t.

Here is what section 6.5 says about C++ types.

> In a SYCL device compiler, the standard C++ fundamental types, including int, short, long, long long int need to be configured so that the device definitions of those types match the host definitions of those types. A device compiler may have this preconfigured so that it can match them based on the definitions of those types on the platform. Or there may be a necessity for a device compiler command-line option to ensure the types are the same.

In other words, defining `cl_*` types is not required for ensuring data type sizes - it must be guaranteed by the SYCL device compiler.

According to my understanding - `cl_*` types are needed only for OpenCL interoperability. For this purpose, `cl_size_t` is not needed as it's not allowed as `size_t` type is not allowed as OpenCL kernel parameter.

@mkinsner, @jbrodman, @rolandschulz, could you follow-up with the spec fix, please?

Reply:
`size_t`  isn't a fundamental type. So I don't think the section 6.5 applies.
I created https://gitlab.devtools.intel.com/SYCL/specbugs/issues/37 . Please add to it with any extra open questions about non-fundamental integer types (and `size_t` in particular).

Reply:
Do not be shy. Report directly on the spec public GitHub instead of the internal Intel GitLab.

Reply:
We will soon. We strive to be as open as possible. Some internal coordination is required.